### PR TITLE
ceph-mds: dump all info of ceph_file_layout, InodeStoreBase, frag_inf…

### PR DESCRIPTION
**Pull request by simon-rock**
…o_t, inode_t

Signed-off-by: simon gao <simon29rock@gmail.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by simon-rock** [comment]
@ukernel  I think we need to retest this.

File: src/common/fs_types.cc [hunk 1]
```diff
@@ -19,6 +19,9 @@ void dump(const ceph_file_layout& l, Formatter *f)
 void dump(const ceph_dir_layout& l, Formatter *f)
 {
   f->dump_unsigned("dir_hash", l.dl_dir_hash);
+  f->dump_unsigned("unused1", l.dl_unused1);
+  f->dump_unsigned("unused2", l.dl_unused2);
+  f->dump_unsigned("unused3", l.dl_unused3);
```
Comment (batrick) [inline src/common/fs_types.cc]: Why do we want to dump this?

File: src/mds/CInode.cc [hunk 2]
```diff
@@ -4027,21 +4027,39 @@ void CInode::decode_import(bufferlist::const_iterator& p,
 
 void InodeStoreBase::dump(Formatter *f) const
 {
+  f->open_object_section("inode");
   inode.dump(f);
+  f->close_section();
   f->dump_string("symlink", symlink);
+
+  f->open_array_section("xattrs");
+  for (const auto &p : xattrs) {
```
Comment (batrick) [inline src/mds/CInode.cc]: use structured binding: `for (const auto& [key, val] : xattrs) {`

File: src/mds/CInode.cc [hunk 3]
```diff
@@ -4027,21 +4027,39 @@ void CInode::decode_import(bufferlist::const_iterator& p,
...
+    f->open_object_section("xattr");
+    f->dump_string("key", p.first);
+    std::string v(p.second.c_str(), p.second.length());
+    f->dump_string("val", v);
```
Comment (batrick) [inline src/mds/CInode.cc]: Why not `f->dump_string("val", val)`?

Comment (simon-rock) [reply src/mds/CInode.cc]: type of xattr is compact_map<string, bufferptr>, os I have to convert bufferptr to string.

File: src/common/fs_types.cc (hunk 1 above)
Comment (simon-rock) [reply src/common/fs_types.cc]: There are three reserved fields, but I know when they will be used. I dump them now and avoid missing this information late.

File: src/mds/CInode.cc (hunk 2 above)
Comment (simon-rock) [reply src/mds/CInode.cc]: ok

File: src/mds/CInode.cc [hunk 4]
```diff
@@ -4027,21 +4027,38 @@ void CInode::decode_import(bufferlist::const_iterator& p,
 
 void InodeStoreBase::dump(Formatter *f) const
 {
+  f->open_object_section("inode");
```
Comment (batrick) [inline src/mds/CInode.cc]: Don't think this separate section is a good thing. It break existing tools and logically it should be part of this InodeStoreBase section.

File: src/mds/CInode.cc (hunk 3 above)
Comment (batrick) [reply src/mds/CInode.cc]: k

File: src/mds/CInode.cc [hunk 5]
```diff
@@ -4027,21 +4027,38 @@ void CInode::decode_import(bufferlist::const_iterator& p,
...
+  f->open_object_section("dirfragtree");
+  dirfragtree.dump(f);
+  f->close_section(); // dirfragtree
+  
   f->open_array_section("old_inodes");
   for (const auto &p : old_inodes) {
     f->open_object_section("old_inode");
     // The key is the last snapid, the first is in the mempool_old_inode
     f->dump_int("last", p.first);
+    f->open_object_section("old_inode_data");
```
Comment (batrick) [inline src/mds/CInode.cc]: this section also looks unnecessary

File: src/mds/CInode.cc (hunk 4 above)
Comment (simon-rock) [reply src/mds/CInode.cc]: @batrick oh. 
I cannot  find the tools that rely to this logic.  I decode the dump file using JSONDecoder.
like this:
```
JSONDecoder::decode_json("inode", inode, obj, true);
=> 
template<class T>
bool JSONDecoder::decode_json(const char *name, T& val, JSONObj *obj, bool mandatory)
```
that need all info of inode_t are in a separate  section when dump it I think.

File: src/mds/CInode.cc (hunk 5 above)
Comment (simon-rock) [reply src/mds/CInode.cc]: @batrick the reason it same to decode inode_t. it's type is old_inode_t.

File: src/mds/CInode.cc (hunk 4 above)
Comment (batrick) [reply src/mds/CInode.cc]: Well, our testing does rely on it:

````
2019-07-20T09:29:57.426 INFO:tasks.cephfs_test_runner:======================================================================
2019-07-20T09:29:57.426 INFO:tasks.cephfs_test_runner:ERROR: test_deep_split (tasks.cephfs.test_fragment.TestFragmentation)
2019-07-20T09:29:57.426 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-07-20T09:29:57.426 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2019-07-20T09:29:57.426 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20190719.231534/qa/tasks/cephfs/test_fragment.py", line 193, in test_deep_split
2019-07-20T09:29:57.426 INFO:tasks.cephfs_test_runner:    frags = self.get_dir_ino("/splitdir")['dirfrags']
2019-07-20T09:29:57.426 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20190719.231534/qa/tasks/cephfs/test_fragment.py", line 25, in get_dir_ino
2019-07-20T09:29:57.427 INFO:tasks.cephfs_test_runner:    if ino['ino'] == dir_inono:
2019-07-20T09:29:57.427 INFO:tasks.cephfs_test_runner:KeyError: 'ino'
````

/ceph/teuthology-archive/pdonnell-2019-07-20_01:58:35-fs-wip-pdonnell-testing-20190719.231534-distro-basic-smithi/4131260/teuthology.log

These extra object sections should be removed.

File: src/mds/CInode.cc (hunk 5 above)
Comment (batrick) [reply src/mds/CInode.cc]: This section too.

**Comment by simon-rock** [comment]
@batrick I have changed.

**Comment by batrick** [comment]
retest this please

**Pull request merged**